src/e_ibmca.c

```diff
@@ -225,540 +226,561 @@ static int ibmca_destroy(ENGINE * e)
-	    || !BIND(ibmca_dso, ica_aes_cfb)
-	    || !BIND(ibmca_dso, ica_des_cfb)
-	    || !BIND(ibmca_dso, ica_3des_cfb)
-	    || !BIND(ibmca_dso, ica_get_functionlist)
+    static int init = 0;
+
+    if (init)                   /* Engine already loaded. */
+        return 1;
+    init++;
+
+    DEBUG_PRINTF(">%s\n", __func__);
+
+    /* Attempt to load libica.so. Needs to be
+     * changed unfortunately because the Ibmca drivers don't have
+     * standard library names that can be platform-translated well. */
+    /* TODO: Work out how to actually map to the names the Ibmca
+     * drivers really use - for now a symbollic link needs to be
+     * created on the host system from libica.so to ica.so on
+     * unix variants. */
+
+    /* WJH XXX check name translation */
+
+    ibmca_dso = dlopen(LIBICA_SHARED_LIB, RTLD_NOW);
+    if (ibmca_dso == NULL) {
+        DEBUG_PRINTF("%s: dlopen(%s) failed\n", __func__, LIBICA_SHARED_LIB);
+        IBMCAerr(IBMCA_F_IBMCA_INIT, IBMCA_R_DSO_FAILURE);
+        goto err;
+    }
+
+    if (!BIND(ibmca_dso, ica_open_adapter)
+        || !BIND(ibmca_dso, ica_close_adapter)
+        || !BIND(ibmca_dso, ica_rsa_mod_expo)
+        || !BIND(ibmca_dso, ica_rsa_crt)
+        || !BIND(ibmca_dso, ica_random_number_generate)
+        || !BIND(ibmca_dso, ica_sha1)
+        || !BIND(ibmca_dso, ica_sha256)
+        || !BIND(ibmca_dso, ica_sha512)
+        || !BIND(ibmca_dso, ica_aes_ecb)
+        || !BIND(ibmca_dso, ica_des_ecb)
+        || !BIND(ibmca_dso, ica_3des_ecb)
+        || !BIND(ibmca_dso, ica_aes_cbc)
+        || !BIND(ibmca_dso, ica_des_cbc)
+        || !BIND(ibmca_dso, ica_3des_cbc)
+        || !BIND(ibmca_dso, ica_aes_ofb)
+        || !BIND(ibmca_dso, ica_des_ofb)
+        || !BIND(ibmca_dso, ica_3des_ofb)
+        || !BIND(ibmca_dso, ica_aes_cfb)
+        || !BIND(ibmca_dso, ica_des_cfb)
+        || !BIND(ibmca_dso, ica_3des_cfb)
+        || !BIND(ibmca_dso, ica_get_functionlist)
 #ifndef OPENSSL_NO_AES_GCM
-	    || !BIND(ibmca_dso, ica_aes_gcm_initialize)
-	    || !BIND(ibmca_dso, ica_aes_gcm_intermediate)
-	    || !BIND(ibmca_dso, ica_aes_gcm_last)
+        || !BIND(ibmca_dso, ica_aes_gcm_initialize)
+        || !BIND(ibmca_dso, ica_aes_gcm_intermediate)
+        || !BIND(ibmca_dso, ica_aes_gcm_last)
 #endif
-	   ) {
-		IBMCAerr(IBMCA_F_IBMCA_INIT, IBMCA_R_DSO_FAILURE);
-		DEBUG_PRINTF("%s: function bind failed\n", __func__);
-		goto err;
-	}
-
-	// disable fallbacks on Libica
-	if (BIND(ibmca_dso, ica_set_fallback_mode))
-		p_ica_set_fallback_mode(0);
-
-        if(!set_supported_meths(e))
-                goto err;
-
-	if (!get_context(&ibmca_handle)) {
-		IBMCAerr(IBMCA_F_IBMCA_INIT, IBMCA_R_UNIT_FAILURE);
-		goto err;
-	}
-
-	DEBUG_PRINTF("<%s success\n", __func__);
-	return 1;
+        ) {
+        IBMCAerr(IBMCA_F_IBMCA_INIT, IBMCA_R_DSO_FAILURE);
+        DEBUG_PRINTF("%s: function bind failed\n", __func__);
+        goto err;
+    }
+    // disable fallbacks on Libica
```

Comment:
I think this comment style should be changed to 
/* disable fallbacks on Libica */